tests/types/test_schema.py

```diff
@@ -247,33 +250,113 @@ def test_schema_inference_on_dictionary(dict_of_ndarrays):
+def test_all_numpy_dtypes():
+    def test_dtype(nparray, dtype):
+        schema = _infer_schema(nparray)
+        assert schema == Schema([TensorSpec(np.dtype(dtype), (-1,))])
+        spec = schema.inputs[0]
+        recreated_spec = TensorSpec.from_json_dict(**spec.to_dict())
+        assert spec == recreated_spec
+
+    bool_ = ["bool", "bool_", "bool8"]
+    object_ = ["object"]
+    signed_int = [
+        "byte",
+        "int8",
+        "short",
+        "int16",
+        "intc",
+        "int32",
+        "int_",
+        "int",
+        "intp",
+        "int64",
+        "longlong",
+    ]
+    unsigned_int = [
+        "ubyte",
+        "uint8",
+        "ushort",
+        "uint16",
+        "uintc",
+        "uint32",
+        "uint",
+        "uintp",
+        "uint64",
+        "ulonglong",
+    ]
+    floating = ["half", "float16", "single", "float32", "double", "float_", "float64"]
+    complex = [
+        "csingle",
+        "singlecomplex",
+        "complex64",
+        "cdouble",
+        "cfloat",
+        "complex_",
+        "complex128",
+        "clongdouble",
+        "clongfloat",
+        "longcomplex",
+        "complex256",
+    ]
+    bytes_ = ["bytes_", "string_"]
+    str_ = ["str_", "unicode_"]
+    platform_dependent = [
+        "longdouble",
+        "longfloat",
+        "float128",
+    ]
 
-    # test floats
-    for t in [np.float16, np.float32, np.float64]:
-        schema = _infer_schema(np.array([1.1, 2.2, 3.3], dtype=t))
-        assert schema == Schema([TensorSpec(np.dtype(t), (-1,))])
-
-    if hasattr(np, "float128"):
-        schema = _infer_schema(np.array([1.1, 2.2, 3.3], dtype=np.float128))
-        assert schema == Schema([TensorSpec(np.dtype(np.float128), (-1,))])
+    # test boolean
+    for dtype in bool_:
+        test_dtype(np.array([True, False, True], dtype=dtype), dtype)
+        test_dtype(np.array([123, 0, -123], dtype=dtype), dtype)
+
+    # test object
+    for dtype in object_:
+        test_dtype(np.array([True, False, True], dtype=dtype), dtype)
+        test_dtype(np.array([123, 0, -123.544], dtype=dtype), dtype)
+        test_dtype(np.array(["test", "this", "type"], dtype=dtype), dtype)
+        test_dtype(np.array(["test", 123, "type"], dtype=dtype), dtype)
+        test_dtype(np.array(["test", 123, 234 + 543j], dtype=dtype), dtype)
+
+    # test signedInt_
+    for dtype in signed_int:
+        test_dtype(np.array([1, 2, 3, -5], dtype=dtype), dtype)
+
+    # test unsignedInt_
+    for dtype in unsigned_int:
+        test_dtype(np.array([1, 2, 3, 5], dtype=dtype), dtype)
+
+    # test floating
+    for dtype in floating:
+        test_dtype(np.array([1.1, -2.2, 3.3, 5.12], dtype=dtype), dtype)
+
+    # test complex
+    for dtype in complex:
+        test_dtype(np.array([1 + 2j, -2.2 - 3.6j], dtype=dtype), dtype)
+
+    # test bytes_
+    for dtype in bytes_:
+        test_dtype(np.array([bytes([1, 255, 12, 34])], dtype=dtype), dtype)
+
+    # str_
+    for dtype in str_:
```

Comment:
the resulting np array created here will have a dtype "U32" or whatever here anyway, but might be more clear to make it explicit that we're creating a numpy with a dtype of a certain string/byte length, and it will still work e.g
```test_dtype(np.array(["a", "bc", "def"], dtype="str10"), "str")```